x/examples/smart-proxy/main.go

```diff
@@ -0,0 +1,147 @@
+// Copyright 2023 Jigsaw Operations LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"net"
+	"net/http"
+	"os"
+	"os/signal"
+	"strings"
+	"time"
+
+	"github.com/Jigsaw-Code/outline-sdk/transport"
+	"github.com/Jigsaw-Code/outline-sdk/x/config"
+	"github.com/Jigsaw-Code/outline-sdk/x/httpproxy"
+	"github.com/Jigsaw-Code/outline-sdk/x/smart"
+)
+
+var debugLog log.Logger = *log.New(io.Discard, "", 0)
+
+type stringArrayFlagValue []string
+
+func (v *stringArrayFlagValue) String() string {
+	return fmt.Sprint(*v)
+}
+
+func (v *stringArrayFlagValue) Set(value string) error {
+	*v = append(*v, value)
+	return nil
+}
+
+func main() {
+	verboseFlag := flag.Bool("v", false, "Enable debug output")
+	addrFlag := flag.String("localAddr", "localhost:1080", "Local proxy address")
+	configFlag := flag.String("config", "config.json", "Address of the config file")
+	transportFlag := flag.String("transport", "", "The base transport for the connections")
+	var domainsFlag stringArrayFlagValue
+	flag.Var(&domainsFlag, "domain", "The test domains to find strategies.")
+
+	flag.Parse()
+	if *verboseFlag {
+		debugLog = *log.New(os.Stderr, "", log.LstdFlags|log.Lmicroseconds|log.Lshortfile)
+	}
+
+	if len(domainsFlag) == 0 {
+		log.Fatal("Must specify flag --domain")
+	}
+
+	if *configFlag == "" {
+		log.Fatal("Must specify flag --config")
+	}
+
+	finderConfig, err := os.ReadFile(*configFlag)
+	if err != nil {
+		log.Fatalf("Could not read config: %v", err)
+	}
+
+	packetDialer, err := config.NewPacketDialer(*transportFlag)
+	if err != nil {
+		log.Fatalf("Could not create packet dialer: %v", err)
+	}
+	streamDialer, err := config.NewStreamDialer(*transportFlag)
+	if err != nil {
+		log.Fatalf("Could not create stream dialer: %v", err)
+	}
+	if strings.HasPrefix(*transportFlag, "ss:") {
```

Comment:
This might not be the only case that the config contains a `ss:` entry. Users may have pipes as well `split:1|ss://...`. It's ok to do a more sophisticated handling in later PRs, but maybe with a todo comment?

Reply:
I was able to come up with a more robust way to detect whether the dialer can be used in Happy Eyeballs. Please see the comments.